lib/puppet/parser/scope.rb

```diff
@@ -587,7 +587,7 @@ def define_settings(type, params)
     }
   end
 
-  RESERVED_VARIABLE_NAMES = ['trusted', 'facts'].freeze
+  RESERVED_VARIABLE_NAMES = ['trusted', 'facts', 'server_facts'].freeze
```

Comment:
yes, exactly. (adding it to RESERVED_VARIABLE_NAMES makes always be reserved, which we cannot do since this is an opt in).